lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py

```diff
@@ -0,0 +1,217 @@
+    if endpoint.url != module.params['url']:
+        return True
+    return False
+
+
+def _system_state_change(module, endpoint):
+    state = module.params['state']
+    if state == 'absent' and endpoint:
+        return True
+
+    if state == 'present':
+        if endpoint is None:
+            return True
+        return _needs_update(module, endpoint)
+
+    return False
+
+
+def main():
+    argument_spec = openstack_full_argument_spec(
+        service=dict(required=True),
+        interface=dict(required=True, choices=['admin', 'public', 'internal']),
+        url=dict(required=True),
+        region=dict(required=False, default=None),
+        enabled=dict(default=True, type='bool'),
+        state=dict(default='present', choices=['absent', 'present']),
+    )
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(argument_spec,
+                           supports_check_mode=True,
+                           **module_kwargs)
+
+    if not HAS_SHADE:
+        module.fail_json(msg='shade is required for this module')
+    if StrictVersion(shade.__version__) < StrictVersion('1.11.0'):
+        module.fail_json(msg="To utilize this module, the installed version of"
+                             "the shade library MUST be >=1.11.0")
+
+    service_name_or_id = module.params['service']
+    interface = module.params['interface']
+    url = module.params['url']
+    region = module.params['region']
+    enabled = module.params['enabled']
+    state = module.params['state']
+
+    try:
+        cloud = shade.operator_cloud(**module.params)
+
+        service = cloud.get_service(service_name_or_id)
+        if service is None:
+            module.fail_json(msg='Service %s does not exist' % service_name_or_id)
+
+        filters = dict(service_id=service.id, interface=interface)
+        if region is not None:
+            filters['region'] = region
+        endpoints = cloud.search_endpoints(filters=filters)
+
+        if len(endpoints) > 1:
+            module.fail_json(msg='Service %s, interface %s and region %s are '
+                                 'not unique' %
+                                 (service_name_or_id, interface, region))
+        elif len(endpoints) == 1:
+            endpoint = endpoints[0]
+        else:
+            endpoint = None
+
+        if module.check_mode:
+            module.exit_json(changed=_system_state_change(module, endpoint))
+
+        if state == 'present':
+            if endpoint is None:
+                result = cloud.create_endpoint(service_name_or_id=service,
+                                               url=url, interface=interface,
+                                               region=region, enabled=True)
+                endpoint = result[0]
+                changed = True
+            else:
+                if _needs_update(module, endpoint):
+                    endpoint = cloud.update_endpoint(
+                        endpoint.id, url=url, enabled=enabled)
```

Comment:
service_name_or_id, interface, and region should also be passed here since that is all updateable

Reply:
not quite.

This module does not accept the endpoint id, it takes service_name_or_id, interface and region to find the endpoint. If the user specifies another name_or_id, interface or region then a matching endpoint will not be found and a new one will be created.

Reply:
ah i see your point. yea that makes sense. well this looks good to me then. thanks for the changes!